Blog: Google Cloud Account Scraper (#2553)
pulumi/pulumi-hugo: themes/default/content/blog/google-cloud-pulumi-import-account-scraper/index.md

---
title: "Google Cloud: Bulk Importing Resources into Pulumi"

> date: 2023-03-07

Reviewer: ```suggestion
date: 2023-03-16
```

meta_desc: "Learn how to bulk import your Google Cloud resources into Pulumi"
meta_image: "meta.png"
authors:
    - josh-kodroff
tags:
    - google-cloud
    - import
---

> Point and click in the console is great when you're first starting out learning a new cloud or managed service, but it quickly becomes a hindrance when cloud is widely adopted by an organization. The point at which the term "widely adopted" becomes applicable to your situation differs, but at some point in their careers many infrastructure and platform engineers are faced with situation where a large number of critical infrastructure resources were created through "clickops" with no ability to track changes, reproduce environments consistently, and so on. When this happens (and it will probably happen to many of you), it's time to import those resources into infrastructure as code.

Reviewer 2: ```suggestion
Point and click in the console is great when you're first starting out learning a new cloud or managed service, but it quickly becomes a hindrance when cloud infrastructure is widely adopted by an organization. The point at which the term "widely adopted" becomes applicable to your situation differs, but at some point in their careers, many infrastructure and platform engineers are faced with situations where a large number of critical infrastructure resources were created through "click ops" with no ability to track changes, reproduce environments consistently, and so on. When this happens (and it will probably happen to many of you), it's time to import those resources into infrastructure as code.
```

> Point and click in the console is great when you're first starting out learning a new cloud or managed service, but it quickly becomes a hinderance when cloud is widely adopted by an organization. The point at which the term "widely adopted" becomes applicable to your situation differs, but at some point in their careers many infrastructure and platform engineers are faced with situation where a large number of critical infrastructure resources were created through "clickops" with no ability to track changes, reproduce environments consistently, and so on. When this happens (and it will probably happen to many of you), it's time to import those resources into infrastructure as code.

Reviewer 3: ```suggestion
Point and click in the console is great when you're first starting out learning a new cloud or managed service, but it quickly becomes a hindrance when cloud is widely adopted by an organization. The point at which the term "widely adopted" becomes applicable to your situation differs, but at some point in their careers many infrastructure and platform engineers are faced with situation where a large number of critical infrastructure resources were created through "clickops" with no ability to track changes, reproduce environments consistently, and so on. When this happens (and it will probably happen to many of you), it's time to import those resources into infrastructure as code.
```

## Writing an account scraper

In our AWS account scraper, we took the approach of querying each resource type using [boto3](https://pypi.org/project/boto3/), the AWS Python SDK. In contrast, Google Cloud has a tool called [Config Connector](https://cloud.google.com/config-connector/docs/overview) that allows us to query all resources in a Google Cloud project with a single command, and without needing to write the code to query each resource type via the Google Cloud SDK.

Our process for importing from Google Cloud has the following steps:

1. Query resources using Config Connector and output them to a YAML file.
1. In a Python script, take the Config Connector YAML output and transform it into JSON suitable for a bulk `pulumi import` operation.
1. Run the `pulumi import` command and place the generated code in our Pulumi program.

> ### Config connector

Reviewer 2: ```suggestion
### Config Connector
```

> Config Connector is designed to manage Google Cloud resources represented as Kubernetes Custom Resource Definitions (CRDs). We will not be using this capability of Config Connector, but if you are interested in managing Pulumi stacks as CRDs in a GitOps workflow, check out the [Pulumi Kubernetes Operator](https://www.pulumi.com/docs/guides/continuous-delivery/pulumi-kubernetes-operator/). Instead, we'll use Config Connector's [bulk export](https://cloud.google.com/config-connector/docs/how-to/import-export/bulk-export#exporting_an_inventory_with_config-connector) capability to query our Google Cloud environment for resources and output them as Kubernetes manifests, then transform those Kubernetes manifests into JSON suitable for a bulk `pulumi import` operation via a script written in Python.

Reviewer 4: The more accurate term is actually just Custom Resources (CR). CRD defines the schema for the CR, but the resources stored in k8s are just CRs.

        'pulumi_type': 'gcp:compute/instance:Instance',
        'get_id': get_compute_instance_id,
    },
    'ComputeInstanceGroup': {
        'pulumi_type': 'gcp:compute/instanceGroup:InstanceGroup',
        'get_id': get_compute_instance_id,
    },
    # ...
}
```

[The full code](https://github.com/pulumi/pulumi-import-google-cloud-account-scraper) also contains some additional mapping and processing for Google Cloud IAM resources, but it follows the same basic patterns as the mapping for other types.

## Importing to Pulumi

Now that we have our conversion script written, we can run it with the following command:

```bash
python3 main.py -i sample-output.yaml -o pulumi-import.json -p <your-project-id>
```

The command will generate output like the following:

```json
{
  "resources": [
    {
      "type": "gcp:artifactregistry/repository:Repository",
      "name": "gcf-artifacts",
      "id": "europe-west1/gcf-artifacts"
    },
    {
      "type": "gcp:compute/disk:Disk",
      "name": "gke-helloworld-2a71d4f-pvc-1e6aa931-d742-437e-a52b-413999ace2be",
      "id": "us-central1-a/gke-helloworld-2a71d4f-pvc-1e6aa931-d742-437e-a52b-413999ace2be"
    },
// ...
```

> Once we're in our Pulumi program's root, we can instruct `pulumi import` to do a bulk import and output the generated code to a file in the language of our program (in this example, TypeScript, but `pulumi import` will automatically generate code in the correct language).

Reviewer 2: ```suggestion
Once we're in our Pulumi program's root, we can instruct `pulumi import` to do a bulk import and output the generated code to a file in the language of our program (this example includes TypeScript output, but `pulumi import` will automatically generate code in the correct language for your project).
```

```typescript
const gcf_artifacts = new gcp.artifactregistry.Repository("gcf-artifacts", {
    description: "This repository is created and used by Cloud Functions for storing function docker images.",
    format: "DOCKER",
    labels: {
        "goog-managed-by": "cloudfunctions",
    },
    location: "europe-west1",
    project: "your-project-id",
    repositoryId: "gcf-artifacts",
});

const gke_helloworld_2a71d4f_pvc_1e6aa931_d742_437e_a52b_413999ace2be = new gcp.compute.Disk("gke-helloworld-2a71d4f-pvc-1e6aa931-d742-437e-a52b-413999ace2be", {
    description: "{\"kubernetes.io/created-for/pv/name\":\"pvc-1e6aa931-d742-437e-a52b-413999ace2be\",\"kubernetes.io/created-for/pvc/name\":\"wpdev-wordpress\",\"kubernetes.io/created-for/pvc/namespace\":\"default\"}",
    labels: {
        "goog-gke-volume": "",
    },
    name: "gke-helloworld-2a71d4f-pvc-1e6aa931-d742-437e-a52b-413999ace2be",
    physicalBlockSizeBytes: 4096,
    project: "your-project-id",
    size: 10,
    zone: "us-central1-a",
});
```

{{% notes %}}

> You may need to massage some of the generated output for any `*IAMPolicy` resources because deleted principals cannot be contained in IAM policies due to a limitation of the Google Classic provider, but they still contained in the generated code returned by `pulumi import`, which in turn comes from the Google API.

Reviewer: ```suggestion
You may need to massage some of the generated output for any `*IAMPolicy` resources because deleted principals cannot be contained in IAM policies due to a limitation of the Google Classic provider, but they are still contained in the generated code returned by `pulumi import`, which in turn comes from the Google API.
```

{{% /notes %}}

## Next Steps

> Now that we have our Google CLoud resources under Pulumi management, we might want to consider some additional steps to build upon our IaC solution:

Reviewer 3: ```suggestion
Now that we have our Google Cloud resources under Pulumi management, we might want to consider some additional steps to build upon our IaC solution:
```

> 

Author: Having trouble thinking of additional text to flesh this out. Per Slack convo, I think it's ok to ship as-is (this content is almost identical to the AWS version), but I'll add text if I can come up with anything to make the transition smoother.

* **Set up a CI/CD pipeline:** By adding our Pulumi code to a delivery pipeline, we can help ensure that all infrastructure changes go through an automated process. Pulumi has helpful integrations and documentation for many popular tools. See [Continuous Delivery](/docs/guides/continuous-delivery/) in the Pulumi docs for more information.

* **Add policy as code:** By adding policy as code to our IaC pipeline, we can ensure that our infrastructure is in compliance with any security or regulatory requirements before it’s ever provisioned in the cloud, as well as checks on any existing Pulumi resources. For more information on Pulumi’s policy as code capabilities, see [CrossGuard (Policy as Code)](/docs/guides/crossguard/) in the Pulumi docs.

## Conclusion

> We’ve shown how the Google Cloud account scraper combined with `pulumi import` enables organizations to quickly get all of their Google Cloud resources under management with Pulumi. Managing cloud resources with Pulumi allows organizations to utilize modern, automated, code-centric approach, and `pulumi import` makes this adopting Pulumi as painless as possible, even when critical resources have been created via point and click in the console!

Reviewer: ```suggestion
We’ve shown how the Google Cloud account scraper combined with `pulumi import` enables organizations to quickly get all of their Google Cloud resources under management with Pulumi. Managing cloud resources with Pulumi allows organizations to utilize modern, automated, code-centric approach, and `pulumi import` makes adopting Pulumi as painless as possible, even when critical resources have been created via point and click in the console!
```
